### PR TITLE
Feature/aws helper function for tags

### PR DESCRIPTION
**Pull request by wimnat**
##### SUMMARY
Added a helper function for comparing AWS tags.

This is a very common exercise in the modules but no helper has existed.

It also tries to encourage a standard way of dealing with tags.  I.e. add a 'purge_tags' option to the module to decide what to do with tags that are not specified in the play.

Also, updated the functions that create tag dicts from tag lists or vice versa so that there is a default key name for tag key and tag value which can be overriden.  There used to be a little if statement for 'Key' or 'key' and 'Value' or 'value' but this is more generic and allows you to override the name when AWS use something non-standard as is the case for AWS KMS (they use TagKey and TagValue).

A couple of modules needed a small change to cater for the above change.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
module_utils/ec2.py

##### ANSIBLE VERSION
2.4

### CODE REVIEW

**Comment by Pallokala** [comment]
How is this proceeding?

I read the code and it seems good in my eyes. I have been anxiously waiting for the AWS Application Load Balancer support to arrive and this seems to be blocking.

File: lib/ansible/modules/cloud/amazon/ec2_eni_facts.py [hunk 1]
```diff
@@ -79,16 +79,17 @@ def list_ec2_snapshots_boto3(connection, module):
...
-    for network_interfaces in snaked_network_interfaces_result['network_interfaces']:
-        network_interfaces['tag_set'] = boto3_tag_list_to_ansible_dict(network_interfaces['tag_set'])
+    # Modify boto3 tags list to be ansible friendly dict and then camel_case
+    camel_network_interfaces = []
+    for network_interface in network_interfaces_result:
+        network_interface['TagSet'] = boto3_tag_list_to_ansible_dict(network_interface['TagSet'])
+        camel_network_interfaces.append(camel_dict_to_snake_dict(network_interface))
 
-    module.exit_json(**snaked_network_interfaces_result)
+    module.exit_json(security_groups=camel_network_interfaces)
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_eni_facts.py]: What does this have to do with security_groups?

File: lib/ansible/modules/cloud/amazon/ec2_group_facts.py [hunk 2]
```diff
@@ -147,17 +147,13 @@ def main():
...
         module.fail_json(msg=e.message, exception=traceback.format_exc())
 
-    # Turn the boto3 result in to ansible_friendly_snaked_names
-    snaked_security_groups = []
+    # Modify boto3 tags list to be ansible friendly dict and then camel_case
+    camel_security_groups = []
     for security_group in security_groups['SecurityGroups']:
-        snaked_security_groups.append(camel_dict_to_snake_dict(security_group))
+        security_group['Tags'] = boto3_tag_list_to_ansible_dict(security_group['Tags'])
+        camel_security_groups.append(camel_dict_to_snake_dict(security_group))
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_group_facts.py]: Should we have a helper that does `camel_list_to_snake_list` so that I don't have to suggest turning this into a list comprehension every time?

File: lib/ansible/module_utils/ec2.py [hunk 3]
```diff
@@ -620,3 +622,32 @@ def map_complex_type(complex_type, type_map):
...
+    tag_key_value_pairs_to_set = {}
+    tag_keys_to_unset = []
+
+    for key in current_tags_dict.keys():
+        if key not in new_tags_dict and purge_tags:
+            tag_keys_to_unset.append(key)
+
+    # Remove the keys we're going to delete then compare the dicts to see if modification is necessary
+    for key in tag_keys_to_unset:
+        del current_tags_dict[key]
```
Comment (willthames) [inline lib/ansible/module_utils/ec2.py]: Will this affect the contents of the dict in the method that calls this (my understanding is that it does unless you do deepcopy) - this might have unintended consequences. I would just do

```
for key in set(new_tags_dict.keys()) - set(tag_keys_to_unset):
     if new_tags_dict[key] != current_tags_dict.get(key):
          tag_key_value_pairs_to_set[key] = new_tags_dict[key]
```

Comment (wimnat) [reply lib/ansible/module_utils/ec2.py]: I like.  Have changed.

File: lib/ansible/modules/cloud/amazon/ec2_eni_facts.py (hunk 1 above)
Comment (wimnat) [reply lib/ansible/modules/cloud/amazon/ec2_eni_facts.py]: Bugger all.  Fixed.

File: lib/ansible/modules/cloud/amazon/ec2_group_facts.py (hunk 2 above)
Comment (wimnat) [reply lib/ansible/modules/cloud/amazon/ec2_group_facts.py]: Probably.  But perhaps ok to leave for another PR?

Comment (willthames) [reply lib/ansible/modules/cloud/amazon/ec2_group_facts.py]: Good point :)

**Comment by willthames** [comment]
shipit, plenty of good use for this, will update any outstanding PRs to use this too, thanks!

**Comment by s-hertel** [comment]
I have things that will use this too. Looks good.

shipit

**Review by s-hertel — APPROVED** [review]

**Comment by s-hertel** [comment]
Re shipit

@willthames

**Comment by wimnat** [comment]
shipit

**Comment by willthames** [comment]
shipit
